src/Core/components/Function.js

```diff
@@ -1563,138 +1563,184 @@ export default class Function extends InlineComponent {
+            fDefinitions[arrayKey] = {
+              functionType: "formula",
+              formula: globalDependencyValues.formula.tree,
+              variables: globalDependencyValues.variables.map(x => x.tree),
+              nInputs: globalDependencyValues.nInputs,
+              nOutputs: globalDependencyValues.nOutputs,
+              domain: globalDependencyValues.domain,
+              component: arrayKey,
             }
           }
-        } else if (dependencyValues.functionChild.length > 0) {
           return {
-            setValue: {
-              fDefinition: dependencyValues.functionChild[0].stateValues.fDefinition
-            }
+            setValue: { fDefinitions }
           }
-        } else if (dependencyValues.numericalfShadow) {
+        } else if (globalDependencyValues.functionChild.length > 0) {
+
+          let fDefinitions = {};
+          for (let arrayKey of arrayKeys) {
+            fDefinitions[arrayKey] = globalDependencyValues.functionChild[0].stateValues.fDefinitions[arrayKey];
+          }
+          return {
+            setValue: { fDefinitions }
+          }
+        } else if (globalDependencyValues.numericalfShadow) {
           // TODO: ??
+          let fDefinitions = {};
+          for (let arrayKey of arrayKeys) {
+            fDefinitions[arrayKey] = {}
+          }
           return {
-            setValue: { fDefinition: {} }
+            setValue: { fDefinitions }
           }
-
-        } else if (dependencyValues.symbolicfShadow) {
+        } else if (globalDependencyValues.symbolicfShadow) {
           // TODO: ??
+          let fDefinitions = {};
+          for (let arrayKey of arrayKeys) {
+            fDefinitions[arrayKey] = {}
+          }
           return {
-            setValue: { fDefinition: {} }
+            setValue: { fDefinitions }
           }
         } else {
-          return {
-            setValue: {
-              fDefinition: {
-                functionType: "formula",
-                formula: dependencyValues.formula.tree,
-                variables: dependencyValues.variables.map(x => x.tree),
-                nInputs: dependencyValues.nInputs,
-                nOutputs: dependencyValues.nOutputs,
-                domain: dependencyValues.domain,
-              }
+
+          let fDefinitions = {};
+          for (let arrayKey of arrayKeys) {
+            fDefinitions[arrayKey] = {
+              functionType: "formula",
+              formula: globalDependencyValues.formula.tree,
+              variables: globalDependencyValues.variables.map(x => x.tree),
+              nInputs: globalDependencyValues.nInputs,
+              nOutputs: globalDependencyValues.nOutputs,
+              domain: globalDependencyValues.domain,
+              component: arrayKey,
             }
           }
+          return {
+            setValue: { fDefinitions }
+          }
         }
       }
     }
 
+    stateVariableDefinitions.fDefinition = {
```

Comment:
I was thinking of asking if we might want to have an alias for the old approach that would access the first element of the array to shrink this diff a bit, but if this isn't likely to break lots of existing docs I guess it is better to have things just removed rather than deprecated. Then I saw this which appears to kind of be implementing that, was this one place where it was just too painful to do all of the refactoring?

Reply:
This concept of alias only works within the DoenetML core itself; their primary purpose is to allow authors to be able to refer to them using $component.alias.  Since fDefinition isn't public, the alias isn't for authors, but it still allows component developers to refer to it in the 
dependencies of state variables.

However, aliases are not propagated outside the web worker to renderers (or to cypress tests), so that code had to be changed to the new convention.